### PR TITLE
pw_unit_test migration: apps-1 and stub-refactoring

### PR DESCRIPTION
**Pull request by Alami-Amine**
Continuation of migration from nl_test to pw_unit_test, effort first discussed in issue https://github.com/project-chip/connectedhomeip/pull/29342

### In this PR:

1. consolidation and refactoring of stub functions that are used by Unit Tests and are related to Interaction Model and ember.
   - the files with the stubs are `test-interaction-model-api` and `test-ember-api`
    - these stub functions were previously defined in the unit test scripts.
   - decoupling them is necessary to allow us to convert tests one by one (to NOT need to convert all tests at once) or we will get linking errors
   
   - The functions to decouple are in TestReadInteraction.cpp , TestWritenteraction.cpp , TestCommandInteraction.cpp, TestAclAttribute.cpp as well as in `TestPowerSourceCluster` 
   - Functions from TestCommandInteraction.cpp weren't moved yet as they use Asserts specific to the unit-test-framework. So i defined them with a weak link.

2. conversion of  a bunch of Tests in `src/app/tests` (Interaction Layer)

### CODE REVIEW

File: src/app/tests/AppTestContext.cpp [hunk 1]
```diff
@@ -49,6 +49,12 @@ void AppContext::SetUpTestSuite()
 
 void AppContext::TearDownTestSuite()
 {
+    // Adding a DrainAndServiceIO call fixes the teardown for some tests (TestAclAttribute and TestReportScheduler); these were
+    // triggering failures in tests that follow them when running on nRF CI (Zephyr native_posix where all Unit Tests are compiled
+    // into a single file)
+    // TODO: understand this more and solve underlying issue
+    LoopbackMessagingContext::DrainAndServiceIO();
```
Comment (bzbarsky-apple) [inline src/app/tests/AppTestContext.cpp:56]: Is this required as part of this change?  Seems like an unrelated fix that could use very careful review not in the context of a mass-change....

File: src/app/tests/TestAclAttribute.cpp [hunk 2]
```diff
@@ -46,18 +47,13 @@ namespace {
...
-chip::EndpointId kTestEndpointId      = 4;
-
 class TestAccessControlDelegate : public AccessControl::Delegate
 {
 public:
     CHIP_ERROR Check(const SubjectDescriptor & subjectDescriptor, const chip::Access::RequestPath & requestPath,
                      Privilege requestPrivilege) override
     {
-        if (requestPath.cluster == kTestDeniedClusterId2)
+        if (requestPath.cluster == chip::Test::kTestDeniedClusterId2)
```
Comment (bzbarsky-apple) [inline src/app/tests/TestAclAttribute.cpp:56]: Why the extra `chip::`?

File: src/app/tests/TestAclAttribute.cpp [hunk 3]
```diff
@@ -166,12 +147,12 @@ void TestAclAttribute::TestACLDeniedAttribute(nlTestSuite * apSuite, void * apCo
                                    chip::app::ReadClient::InteractionType::Subscribe);
 
         chip::app::AttributePathParams attributePathParams[2];
-        attributePathParams[0].mEndpointId  = kTestEndpointId;
-        attributePathParams[0].mClusterId   = kTestDeniedClusterId1;
+        attributePathParams[0].mEndpointId  = chip::Test::kTestEndpointId;
+        attributePathParams[0].mClusterId   = chip::Test::kTestDeniedClusterId1;
```
Comment (bzbarsky-apple) [inline src/app/tests/TestAclAttribute.cpp:151]: Why alll the extra `chip::`?  This applies to a bunch of places here.

File: src/app/tests/TestReadInteraction.cpp [hunk 4]
```diff
@@ -301,53 +303,54 @@ using ReadHandlerNode     = chip::app::reporting::ReportScheduler::ReadHandlerNo
...
-    }
-
-    return false;
-}
-
-bool IsDeviceTypeOnEndpoint(DeviceTypeId deviceType, EndpointId endpoint)
-{
-    return false;
-}
+// CHIP_ERROR ReadSingleClusterData(const Access::SubjectDescriptor & aSubjectDescriptor, bool aIsFabricFiltered,
```
Comment (bzbarsky-apple) [inline src/app/tests/TestReadInteraction.cpp:306]: I assume the commented-out bits will go away before this is un-drafted?

File: src/app/tests/ember-test-compatibility.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+// was previously in TestWriteInteraction.cpp
+const EmberAfAttributeMetadata * GetAttributeMetadata(const ConcreteAttributePath & aConcreteClusterPath)
+{
+    // Note: This test does not make use of the real attribute metadata.
+    static EmberAfAttributeMetadata stub = { .defaultValue = EmberAfDefaultOrMinMaxAttributeValue(uint32_t(0)) };
+    return &stub;
+}
+
+// was previously in TestWriteInteraction.cpp
+CHIP_ERROR WriteSingleClusterData(const Access::SubjectDescriptor & aSubjectDescriptor, const ConcreteDataAttributePath & aPath,
```
Comment (bzbarsky-apple) [inline src/app/tests/ember-test-compatibility.cpp:27]: This has nothing to do with "ember".  Same for a lot of the other functions being added here.

File: src/app/tests/AppTestContext.cpp (hunk 1 above)
Comment (Alami-Amine) [reply src/app/tests/AppTestContext.cpp:56]: well, I only need it in order to be able to run `TestAclAttribute.cpp` and `TestTestReportScheduler.cpp` before the other tests. 
I have just put it in a seperate PR: https://github.com/project-chip/connectedhomeip/pull/33654

File: src/app/tests/TestAclAttribute.cpp (hunk 3 above)
Comment (Alami-Amine) [reply src/app/tests/TestAclAttribute.cpp:151]: ok will review them all

File: src/app/tests/ember-test-compatibility.cpp (hunk 5 above)
Comment (Alami-Amine) [reply src/app/tests/ember-test-compatibility.cpp:27]: if I understood correctly, the original form of this function (the one defined in `ember-compatibility-functions.cpp`) interacts directly with ember layer. and this is a stub definition to override it, no?
 what do you suggest I call it?

File: src/app/tests/TestAclAttribute.cpp (hunk 2 above)
Comment (Alami-Amine) [reply src/app/tests/TestAclAttribute.cpp:56]: because the pigweed namespace is also called `Test`... putting only Test chooses `pw::unit_test::internal::Test`

File: src/app/tests/TestAclAttribute.cpp (hunk 3 above)
Comment (Alami-Amine) [reply src/app/tests/TestAclAttribute.cpp:151]: because the pigweed namespace is also called `Test`... putting only Test chooses `pw::unit_test::internal::Test`

File: src/app/tests/ember-test-compatibility.cpp (hunk 5 above)
Comment (bzbarsky-apple) [reply src/app/tests/ember-test-compatibility.cpp:27]: I suggest you call this `test-interaction-model-api.cpp` or something like that.  `WriteSingleClusterData` is a backend-agnostic API that the IM uses.  The implementation in ember-compatibility implements `WriteSingleClusterData` on top of ember bits, but your implementation has nothing to do with ember....

Comment (Alami-Amine) [reply src/app/tests/ember-test-compatibility.cpp:29]: FYI, it was suggested by @bzbarsky-apple to create another file for the ember specific bit: since I just added a new stub `emberAfGetClusterServerEndpointIndex()` which is in fact more ember related

File: src/app/tests/TestAttributeValueEncoder.cpp [hunk 6]
```diff
@@ -18,10 +18,15 @@
 
 /**
  *    @file
- *      This file implements unit tests for CommandPathParams
+ *      This file implements unit tests for AttributeValueEncoder
```
Comment (andy31415) [inline src/app/tests/TestAttributeValueEncoder.cpp:21]: nit: instead of fixing I would consider dropping this entire comment block: it often has negative benefit:

- it is obvious `a file named testAttributeValueEncoder obviously tessts the AttributeValueEncoder`
- tests are copied and pasted for license and header include boilerplate ... and then this comment becomes copy and pasted and sometimes not updated, making us spend time to update it.

File: src/app/tests/TestBindingTable.cpp [hunk 7]
```diff
@@ -15,122 +15,122 @@
...
 
-void TestEmptyBindingTable(nlTestSuite * aSuite, void * aContext)
+void VerifyTableSame(BindingTable & table, const std::vector<EmberBindingTableEntry> & expected)
+{
+    EXPECT_EQ(table.Size(), expected.size());
+    auto iter1 = table.begin();
+    auto iter2 = expected.begin();
+    while (iter2 != expected.end())
+    {
+        EXPECT_NE(iter1, table.end());
```
Comment (andy31415) [inline src/app/tests/TestBindingTable.cpp:37]: if we use EXPECT then processing continues and the line below will deference invalid data probably.

Should we maybe do ASSERT_EQ on line 32 above and then not need this?

File: src/app/tests/TestReadInteraction.cpp [hunk 8]
```diff
@@ -54,14 +56,14 @@ uint8_t gDebugEventBuffer[128];
...
-chip::EventId kTestEventIdCritical      = chip::Test::MockEventId(2);
-uint8_t kTestFieldValue1                = 1;
+chip::ClusterId kTestClusterId        = 6;
+chip::ClusterId kTestEventClusterId   = chip::Test::MockClusterId(1);
+chip::ClusterId kInvalidTestClusterId = 7;
+chip::EndpointId kTestEndpointId      = 1;
+chip::EndpointId kTestEventEndpointId = chip::Test::kMockEndpoint1;
+chip::EventId kTestEventIdDebug       = chip::Test::MockEventId(1);
+chip::EventId kTestEventIdCritical    = chip::Test::MockEventId(2);
+// uint8_t kTestFieldValue1                = 1;
```
Comment (andy31415) [inline src/app/tests/TestReadInteraction.cpp:66]: shoudl we delete things instead of commenting them out? Where was this used before?

File: src/app/tests/TestTimeSyncDataProvider.cpp [hunk 9]
```diff
@@ -201,98 +207,65 @@ void TestDSTOffset(nlTestSuite * inSuite, void * inContext)
...
-    VerifyOrReturnError(CHIP_NO_ERROR == chip::Platform::MemoryInit(), FAILURE);
-    return SUCCESS;
-}
-
-int TestTearDown(void * inContext)
-{
-    chip::Platform::MemoryShutdown();
-    return SUCCESS;
+    EXPECT_EQ(CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND, timeSyncDataProv.LoadDSTOffset(dstObj));
+    EXPECT_FALSE(dstObj.dstOffsetList.begin());
```
Comment (andy31415) [inline src/app/tests/TestTimeSyncDataProvider.cpp:267]: this reads awakward. Does `EXPECT_TRUE(dstObj.dstOffsetList.empty())` work ? Or maybe:

`EXPECT_EQ(dstObj.dstOffsetList.begin(), dstObj.dstOffsetList.end())`

or is this really wanting to check for nullability? maybe then we should compare with nullptr, however a compare by converting an iterator to a bool feels off.

File: src/app/tests/test-ember-api.cpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <app/tests/test-ember-api.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/basic-types.h>
+#include <app/util/mock/Constants.h>
+#include <app/util/mock/Functions.h>
+
+chip::EndpointId chip::Test::numEndpoints = 0;
+
+// was previously in TestPowerSourceCluster.cpp
```
Comment (andy31415) [inline src/app/tests/test-ember-api.cpp:25]: I think the `previously` may lose context as time goes by. How about:
```suggestion
// Used by code in TestPowerSourceCluster.cpp (and generally things using mock ember functions may need this).
```

File: src/app/tests/test-ember-api.h [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/util/basic-types.h>
+
+/// test-ember-api was created to consolidate and centralize stub functions that are related to ember and are run for unit-tests
+/// These were previously defined in the unit test-scripts themselves.
```
Comment (andy31415) [inline src/app/tests/test-ember-api.h:19]: I would rephrase to lose `previously` to make the comment not reference potentially very old code. We have context now but it will be lost in months/years.

File: src/app/tests/test-interaction-model-api.cpp [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+#include <lib/core/CHIPCore.h>
+#include <messaging/ReliableMessageContext.h>
+
+namespace chip {
+uint8_t Test::attributeDataTLV[CHIP_CONFIG_DEFAULT_UDP_MTU_SIZE];
+size_t Test::attributeDataTLVLen = 0;
+
+namespace app {
+
+// was previously in TestWriteInteraction.cpp
```
Comment (andy31415) [inline src/app/tests/test-interaction-model-api.cpp:35]: same here and maybe more instances. I will stop commenting on this particular thing.

File: src/app/tests/test-interaction-model-api.cpp [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+// was previously in TestReadInteraction.cpp
+bool IsClusterDataVersionEqual(const ConcreteClusterPath & aConcreteClusterPath, DataVersion aRequiredVersion)
+{
+    if (Test::kTestDataVersion1 == aRequiredVersion)
+    {
+        return true;
+    }
+
+    return false;
```
Comment (andy31415) [inline src/app/tests/test-interaction-model-api.cpp:99]:
```suggestion
    return (Test::kTestDataVersion1 == aRequiredVersion);
```

**Review by andy31415 — APPROVED** [review]

File: src/app/tests/TestStatusResponseMessage.cpp [hunk 14]
```diff
@@ -22,130 +22,92 @@
  *
  */
 
+#include <gtest/gtest.h>
```
Comment (feasel0) [inline src/app/tests/TestStatusResponseMessage.cpp:25]:
```suggestion
#include <pw_unit_test/framework.h>
```

File: src/app/tests/TestStatusIB.cpp [hunk 15]
```diff
@@ -15,193 +15,150 @@
  *    limitations under the License.
  */
 
+#include <gtest/gtest.h>
```
Comment (feasel0) [inline src/app/tests/TestStatusIB.cpp:18]:
```suggestion
#include <pw_unit_test/framework.h>
```

File: src/app/tests/TestPendingResponseTrackerImpl.cpp [hunk 16]
```diff
@@ -15,121 +15,99 @@
...
 
-    NL_TEST_ASSERT(inSuite, std::numeric_limits<uint16_t>::max() == pendingResponseTracker.Count());
+    EXPECT_EQ(std::numeric_limits<uint16_t>::max(), pendingResponseTracker.Count());
 
     for (uint16_t commandRef = 0; commandRef < std::numeric_limits<uint16_t>::max(); commandRef++)
     {
-        NL_TEST_ASSERT(inSuite, CHIP_NO_ERROR == pendingResponseTracker.Remove(commandRef));
-        NL_TEST_ASSERT(inSuite, false == pendingResponseTracker.IsTracked(commandRef));
+        EXPECT_EQ(CHIP_NO_ERROR, pendingResponseTracker.Remove(commandRef));
+        EXPECT_EQ(false, pendingResponseTracker.IsTracked(commandRef));
```
Comment (feasel0) [inline src/app/tests/TestPendingResponseTrackerImpl.cpp:44]:
```suggestion
        EXPECT_FALSE(pendingResponseTracker.IsTracked(commandRef));
```

File: src/app/tests/TestBuilderParser.cpp [hunk 17]
```diff
@@ -16,6 +16,8 @@
  *    limitations under the License.
  */
 
+#include <gtest/gtest.h>
```
Comment (feasel0) [inline src/app/tests/TestBuilderParser.cpp:19]:
```suggestion
#include <pw_unit_test/framework.h>
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/tests/TestMessageDef.cpp [hunk 18]
```diff
@@ -22,6 +22,8 @@
  *
  */
 
+#include <gtest/gtest.h>
```
Comment (feasel0) [inline src/app/tests/TestMessageDef.cpp:25]:
```suggestion
#include <pw_unit_test/framework.h>
```

**Review by feasel0 — APPROVED** [review]

File: src/app/tests/TestAttributeValueEncoder.cpp (hunk 6 above)
Comment (Alami-Amine) [reply src/app/tests/TestAttributeValueEncoder.cpp:21]: totally Agree, this same exact comment is already in half the tests

File: src/app/tests/TestReadInteraction.cpp (hunk 8 above)
Comment (Alami-Amine) [reply src/app/tests/TestReadInteraction.cpp:66]: i will delete it, this was here before, and now I moved it to `test-interaction-model-api` where it is used by the function `ReadSingleClusterData`

File: src/app/tests/TestStatusResponseMessage.cpp (hunk 14 above)
Comment (Alami-Amine) [reply src/app/tests/TestStatusResponseMessage.cpp:25]: thanks, I also added `#include <lib/core/StringBuilderAdapters.h>` as a bonus

File: src/app/tests/TestPendingResponseTrackerImpl.cpp (hunk 16 above)
Comment (Alami-Amine) [reply src/app/tests/TestPendingResponseTrackerImpl.cpp:44]: good catch

File: src/app/tests/TestBindingTable.cpp (hunk 7 above)
Comment (Alami-Amine) [reply src/app/tests/TestBindingTable.cpp:37]: yes sure, assuming that iter1 will not have an invalid index?

File: src/app/tests/TestTimeSyncDataProvider.cpp (hunk 9 above)
Comment (Alami-Amine) [reply src/app/tests/TestTimeSyncDataProvider.cpp:267]: yes emtpy works fine, done!

File: src/app/tests/test-ember-api.cpp (hunk 10 above)
Comment (Alami-Amine) [reply src/app/tests/test-ember-api.cpp:25]: done
